### ISSUE TITLE
Lack of entropy during boot hangs system

### ISSUE THREAD

**Issue by BenWiederhake** [SerenityOS/serenity#2871]
tl;dr: Lack of entropy means that processes cannot be created.

Process creation should instead only require weak entropy.

## Log lines

``` 
[Kernel]: CPU[0]: Supported features: nx pae pge
[Kernel]: CPU[0]: No RDRAND support detected, randomness will be poor
[Kernel]: Installing Unhandled Handlers
[Kernel]: MM: boot_pdpt @ P0x00109000
[Kernel]: MM: boot_pd0 @ P0x0010a000
[Kernel]: MM: boot_pd3 @ P0x0010b000
<SNIP FOR BREVITY>
[colonel(0:0)]: Scheduler[0]: idle loop running
[init_stage2(1:1)]: BW: about to mount VFS
[init_stage2(1:1)]: VFS: mounted root from Ext2FS (device:3,0 (PATADiskDevice))
[init_stage2(1:1)]: Loading kernel symbol table...
[init_stage2(1:1)]: BW: done loading
[init_stage2(1:1)]: BW: done set graphical
[init_stage2(1:1)]: BW: done parsing commandline
```

I inserted those "BW" lines to narrow it down, and it consistently hangs on this line:

```C++
// init.cpp:349
Process::create_user_process(thread, userspace_init, (uid_t)0, (gid_t)0, (pid_t)0, error, {}, {}, tty0);
```

## I built it correctly

When I follow the build instructions in [`BuildInstructions.md`](https://github.com/SerenityOS/serenity/blob/master/Documentation/BuildInstructions.md) quite verbatim, serenity hangs.

If you'd like to know [yes](https://github.com/BenWiederhake/serenity-hang/blob/master/serenity-from-scratch.sh) I have followed the [entire build process from scratch](https://github.com/BenWiederhake/serenity-hang/blob/master/full-build-from-scratch-1595639004.log).

## I waited long enough

After two minutes, still nothing happens.

## Impatience gets rewarded!?

The following things seem to get the Kernel "unstuck" though:
- Clicking about 20 times
- Typing about 100 keys
- Moving the mouse around for about 5 seconds
(These numbers are just guesstimates to give you a feel. Please don't actually count the clicks/keys/seconds.)

## Randomness?

```
(04:51:42) Ben: When scrolling through the commit logs it looked like Andreas is using an RDRAND processor, and other processors are not tested as thoroughly.
(04:51:59) Ben: My processor doesn't have RDRAND, so maybe that has something to do with it.
(04:52:35) Ben: That would leave the question why SystemServer hangs, because why should it need randomness?
(04:53:28) CxByte: I guess we don't use a shitty algorithm when there's no entropy then...get_good_random_bytes() does a wait_for_entropy()
(04:54:02) CxByte: need randomness: Don't we give some sort of random value to every process? lemme check
(04:56:00) CxByte: yep, do_exec() -> generate_auxiliary_vector() -> get_good_random_bytes()
(04:56:34) Ben: That would explain why SystemServer "needs" entropy.
```
(Thanks, CxByte!)

## Potential fixes

I see two ways:
1. Try even harder to get good entropy on a "shut-in" machine without RDRAND and with too few naturally-occurring interrupts during boot. (But then we'll end up with the same problem sooner or later. Also, the Kernel should *show* that this is what's going on.)
1. Remove the requirement of giving good entropy to SystemServer. (But then it will hang on process 2, whatever that is.)
1. When waiting for new entropy to trickle in, use the time to cause new entropy, e.g. by harddrive reads. (But that's a lot of work to implement, and is frankly hard to argue as correct.)
1. Remove the requirement of giving good entropy to new processes.

## The fix

So … `generate_auxiliary_vector()` should probably call `get_fast_random_bytes()`.

However, I don't know for sure where this entropy is used, or why it was originally decided that strong entropy is needed here, so I don't want to change it yet.

## Workaround

Wiggle your mouse.

Stop laughing! I'm a keyboard guy, so this *is* an entirely separate step for me! :)

## Why so elaborate

This was an unintentional rubberduck. I started writing this thinking that Serenity/SystemServer/qemu are behaving completely unpredictably. Then I thought it's something to do with Kernel cmdline parsing. Then I thought it's something to do with interrupts. Then I thought it's maybe a qemu bug. Then I thought maybe it's a "promptless" prompt. Then I thought it's a button that I accidentally press when clicking blindly. Then I thought it's something that increases with each interrupt … Oh god it's the entropy bug all over again.

I hope you enjoyed the ride. I certainly did! :^)

**Comment by BenWiederhake** [comment]
Linux doesn't adhere to the "securely generated random bytes" requirement for `AT_RANDOM`:

```
$ { for _ in $(seq 40) ; do LD_SHOW_AUXV=1 /bin/true ; done } 2>&1 | grep RANDOM
AT_RANDOM:            0x7ffe951c1479
AT_RANDOM:            0x7ffc00b10f79
AT_RANDOM:            0x7ffdb3dfdb49
AT_RANDOM:            0x7ffd54b4d429
AT_RANDOM:            0x7ffdc2e788c9
AT_RANDOM:            0x7ffc5340e3e9
AT_RANDOM:            0x7ffddb769b99
AT_RANDOM:            0x7ffed848e0b9
AT_RANDOM:            0x7fffb4704c19
AT_RANDOM:            0x7ffc1f2676c9
AT_RANDOM:            0x7fffcd19ea89
AT_RANDOM:            0x7fffb00dbba9
AT_RANDOM:            0x7ffd2084bdf9
AT_RANDOM:            0x7ffc9aa55e89
AT_RANDOM:            0x7fff2adb66a9
AT_RANDOM:            0x7ffd3f725159
AT_RANDOM:            0x7ffcd273e9b9
AT_RANDOM:            0x7ffec2b96929
AT_RANDOM:            0x7ffce3629999
AT_RANDOM:            0x7ffc6aa5a3f9
AT_RANDOM:            0x7fff21b65f89
AT_RANDOM:            0x7ffd5d6b00e9
AT_RANDOM:            0x7ffe29c610a9
AT_RANDOM:            0x7ffce53211c9
AT_RANDOM:            0x7ffc1c024da9
AT_RANDOM:            0x7ffc8ea2d039
AT_RANDOM:            0x7ffc7621f269
AT_RANDOM:            0x7ffe86804c29
AT_RANDOM:            0x7fff46e394c9
AT_RANDOM:            0x7ffcd444de19
AT_RANDOM:            0x7ffcae15d6c9
AT_RANDOM:            0x7fff43571009
AT_RANDOM:            0x7fff0c87fa39
AT_RANDOM:            0x7fff59a768f9
AT_RANDOM:            0x7ffeeb346b69
AT_RANDOM:            0x7ffebf8a7139
AT_RANDOM:            0x7fff04585bd9
AT_RANDOM:            0x7ffe1358ffa9
AT_RANDOM:            0x7ffc0a866269
AT_RANDOM:            0x7ffdad23a7e9
```

So I believe it's not too bad if Serenity falls back to somewhat-insecure behavior if entropy is abysmally low.
